agent/rpc/peering/replication.go

```diff
@@ -282,8 +288,136 @@ func (s *Service) handleUpsertService(
+		// A single node check will be attached to all service instances of a node, so this
+		// deduplication prevents issuing multiple deregistrations for a single check.
+		deletedNodeChecks = make(map[nodeCheckTuple]struct{})
+	)
+	for _, csn := range storedInstances {
+		if _, ok := snap.Nodes[csn.Node.ID]; !ok {
+			unusedNodes[csn.Node.ID] = struct{}{}
+
+			// Since the node is not in the snapshot we can know the associated service
+			// instance is not in the snapshot either, since a service instance can't
+			// exist without a node.
+			// This will also delete all service checks.
+			err := s.Backend.Apply().CatalogDeregister(&structs.DeregisterRequest{
+				Node:           csn.Node.Node,
+				ServiceID:      csn.Service.ID,
+				EnterpriseMeta: csn.Service.EnterpriseMeta,
+				PeerName:       peerName,
+			})
+			if err != nil {
+				return fmt.Errorf("failed to deregister service: %w", err)
+			}
+
+			// We can't know if a node check was deleted from the exporting cluster
+			// (but not the node itself) if the node wasn't in the snapshot,
+			// so we do not loop over checks here.
+			// If the unusedNode gets deleted below that will also delete node checks.
+			continue
+		}
+
+		// Delete the service instance if not in the snapshot.
+		sid := csn.Service.CompoundServiceID()
+		if _, ok := snap.Nodes[csn.Node.ID].Services[sid]; !ok {
+			err := s.Backend.Apply().CatalogDeregister(&structs.DeregisterRequest{
+				Node:           csn.Node.Node,
+				ServiceID:      csn.Service.ID,
+				EnterpriseMeta: csn.Service.EnterpriseMeta,
+				PeerName:       peerName,
+			})
+			if err != nil {
+				return fmt.Errorf("failed to deregister service: %w", err)
+			}
+
+			// When a service is deleted all associated checks also get deleted as a side effect.
+			continue
+		}
+
+		// Reconcile checks.
+		for _, chk := range csn.Checks {
+			if _, ok := snap.Nodes[csn.Node.ID].Services[sid].Checks[chk.CheckID]; !ok {
+				// Checks without a ServiceID are node checks.
+				// If the node exists but the check does not then the check was deleted.
+				if chk.ServiceID == "" {
+					// Deduplicate node checks to avoid deregistering a check multiple times.
+					tuple := nodeCheckTuple{
+						checkID: chk.CheckID,
+						node:    chk.Node,
+					}
+					deletedNodeChecks[tuple] = struct{}{}
+					continue
+				}
+
+				// If the check isn't a node check then it's a service check.
+				// Service checks that were not present can be deleted immediately because
+				// checks for a given service ID will only be attached to a single CheckServiceNode.
+				err := s.Backend.Apply().CatalogDeregister(&structs.DeregisterRequest{
+					Node:           chk.Node,
+					CheckID:        chk.CheckID,
+					EnterpriseMeta: chk.EnterpriseMeta,
+					PeerName:       peerName,
+				})
+				if err != nil {
+					return fmt.Errorf("failed to deregister check: %w", err)
+				}
+			}
+		}
+	}
 
+	// Delete all deduplicated node checks.
+	for chk := range deletedNodeChecks {
```

Comment:
I don't think you can do this, because they'd be shared by other services right? Or are you assuming that if it vanished from THIS service on the node that implies it vanished from the node and is thus safe for erasure?

Reply:
> Or are you assuming that if it vanished from THIS service on the node that implies it vanished from the node and is thus safe for erasure?

Yep that's right. Is there a case where that assumption wouldn't hold?

Reply:
I'm going to rescind this comment after I read through the rest of this.